src/lumigo_tracer/lambda_tracer/lambda_reporter.py

```diff
@@ -168,36 +182,135 @@ def report_json(
+
+    First we try to take all the spans and then we apply the smart span selection.
+
+    The smart span selection has 2 parts:
+    1. We order the spans by FUNCTION_SPAN, ERROR_HTTP_SPAN, ENRICHMENT_SPAN, HTTP_SPAN.
+    2. We take all the spans metadata, We take the full spans. We do this until reach the max_size.
+    """
+    request_size_limit = max_error_size if any(map(is_span_has_error, msgs)) else max_size
 
     if not prune_size_flag or (
         len(msgs) < NUMBER_OF_SPANS_IN_REPORT_OPTIMIZATION
-        and _get_event_base64_size(msgs) < max_size  # noqa
+        and _get_event_base64_size(msgs) < request_size_limit  # noqa
     ):
-        return aws_dump(msgs)[:max_size]
-
-    end_span = msgs[-1]
-    ordered_spans = sorted(msgs[:-1], key=is_span_has_error, reverse=True)
+        return aws_dump(msgs)[:request_size_limit]
 
-    spans_to_send: list = [end_span]  # type: ignore[type-arg]
-    current_size = _get_event_base64_size(end_span)
-    too_big_spans = 0
-    for span in ordered_spans:
+    current_size = 0
+    spans_to_send: List[dict] = []  # type: ignore[type-arg]
+    for index, span in enumerate(msgs):
         span_size = _get_event_base64_size(span)
-        if current_size + span_size < max_size:
-            spans_to_send.append(span)
-            current_size += span_size
-        else:
-            # This is an optimization step. If the spans are too big, don't try to send them.
-            too_big_spans += 1
-            if too_big_spans == too_big_spans_threshold:
-                break
-    return aws_dump(spans_to_send)[:max_size]
+        if current_size + span_size > max_size:
+            break
+
+        spans_to_send.append(span)
+        current_size += span_size
+
+    if len(spans_to_send) < len(msgs):
+        with lumigo_safe_execute("create_request_body: smart span selection"):
+            get_logger().info("Starting smart span selection")
+            # If we didn't send all the spans, we need to apply the smart span selection.
+            ordered_spans = sorted(msgs, key=get_span_priority)
+            current_size = 0
+            too_big_spans = 0
+            spans_to_send_sizes = {}
+            spans_to_send_dict = {}
+
+            # Take only spans metadata
+            for index, span in enumerate(ordered_spans):
+                spans_to_send_sizes[index] = 0
+                span_metadata = get_span_metadata(span)
+                if span_metadata == {}:
+                    continue
+                span_metadata_size = _get_event_base64_size(span_metadata)
+
+                if span.get("type") == FUNCTION_TYPE:
+                    # We always want to at least send the function span
+                    spans_to_send_dict[index] = span_metadata
+                    spans_to_send_sizes[index] = span_metadata_size
+                    current_size += span_metadata_size
+                elif current_size + span_metadata_size < max_size:
```

Comment:
```suggestion
                if span.get("type") == FUNCTION_TYPE or current_size + span_metadata_size < max_size:
                    # We always want to at least send the function span
                    spans_to_send_dict[index] = span_metadata
                    spans_to_send_sizes[index] = span_metadata_size
                    current_size += span_metadata_size
```
right? did i missread?